flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java

```diff
@@ -671,13 +675,20 @@ public void setNumBuffers(int numBuffers) {
 
             currentPoolSize = Math.min(numBuffers, maxNumberOfMemorySegments);
 
-            // reset overdraft buffers
+            // If pool size increases, try to convert overdraft buffer to ordinary buffer.
             while (numberOfRequestedOverdraftMemorySegments > 0
                     && numberOfRequestedMemorySegments < currentPoolSize) {
                 numberOfRequestedOverdraftMemorySegments--;
                 numberOfRequestedMemorySegments++;
             }
 
+            // If pool size decreases, try to convert ordinary buffer to overdraft buffer.
+            while (numberOfRequestedMemorySegments > currentPoolSize
+                    && numberOfRequestedOverdraftMemorySegments < maxOverdraftBuffersPerGate) {
+                numberOfRequestedMemorySegments--;
+                numberOfRequestedOverdraftMemorySegments++;
+            }
```

Comment:
Hi @reswqa , thanks for the fix.

I'm curious, what's wrong with just leaving `numberOfRequestedMemorySegments > currentPoolSize`? Count extra buffers to the `overdraft buffer`.

```suggestion
            while (numberOfRequestedMemorySegments > currentPoolSize) {
                numberOfRequestedMemorySegments--;
                numberOfRequestedOverdraftMemorySegments++;
            }
```

Reply:
Thanks for the clarification.

> When numberOfRequestedMemorySegments <= poolSize, all buffers are ordinary buffer.
> When numberOfRequestedMemorySegments > poolSize, the `ordinary buffer size = poolSize`, and `the overdraft buffer size = numberOfRequestedMemorySegments - poolSize`

From the discussion of FLINK-31610, this restriction(`numberOfRequestedOverdraftMemorySegments <= maxOverdraftBuffersPerGate`) will be broken in FLINK-31764, right?

If so, I prefer FLINK-31764 only refactor the code, not change the overdraft strategy or behavior, and we change the strategy in this PR, WDYT?

Please correct my if my understanding is wrong, thanks!

Reply:
> From the discussion of [FLINK-31610](https://issues.apache.org/jira/browse/FLINK-31610), this restriction(numberOfRequestedOverdraftMemorySegments <= maxOverdraftBuffersPerGate) will be broken in [FLINK-31764](https://issues.apache.org/jira/browse/FLINK-31764), right?

Yep, you are absolutely right.

> If so, I prefer [FLINK-31764](https://issues.apache.org/jira/browse/FLINK-31764) only refactor the code, not change the overdraft strategy or behavior, and we change the strategy in this PR, WDYT?

Change the semantic/behavior in this PR also make sense to me. Let's update FLINK-31764 to only simply remove the redundant field.